substrate/primitives/core/src/crypto_bytes.rs

```diff
@@ -0,0 +1,233 @@
+	hash::{H256, H512},
+};
+
+use codec::{Decode, Encode, MaxEncodedLen};
+use core::marker::PhantomData;
+use scale_info::TypeInfo;
+use sp_runtime_interface::pass_by::{self, PassBy, PassByInner};
+
+/// Generic byte array.
+///
+/// The type is generic over a constant length `N` and a "tag" `T` which
+/// can be used to specialize the byte array without using newtypes.
+#[derive(Encode, Decode, MaxEncodedLen)]
+pub struct CryptoBytes<const N: usize, T = ()>(pub [u8; N], PhantomData<fn() -> T>);
+
+impl<const N: usize, T> Copy for CryptoBytes<N, T> {}
+
+impl<const N: usize, T> Clone for CryptoBytes<N, T> {
+	fn clone(&self) -> Self {
+		Self(self.0, PhantomData)
+	}
+}
+
+impl<const N: usize, T> TypeInfo for CryptoBytes<N, T> {
+	type Identity = [u8; N];
+
+	fn type_info() -> scale_info::Type {
+		Self::Identity::type_info()
+	}
+}
+
+impl<const N: usize, T> PartialOrd for CryptoBytes<N, T> {
+	fn partial_cmp(&self, other: &Self) -> Option<core::cmp::Ordering> {
+		self.0.partial_cmp(&other.0)
+	}
+}
+
+impl<const N: usize, T> Ord for CryptoBytes<N, T> {
+	fn cmp(&self, other: &Self) -> core::cmp::Ordering {
+		self.0.cmp(&other.0)
+	}
+}
+
+impl<const N: usize, T> PartialEq for CryptoBytes<N, T> {
+	fn eq(&self, other: &Self) -> bool {
+		self.0.eq(&other.0)
+	}
+}
+
+impl<const N: usize, T> core::hash::Hash for CryptoBytes<N, T> {
+	fn hash<H: scale_info::prelude::hash::Hasher>(&self, state: &mut H) {
+		self.0.hash(state)
+	}
+}
+
+impl<const N: usize, T> Eq for CryptoBytes<N, T> {}
+
+impl<const N: usize, T> Default for CryptoBytes<N, T> {
+	fn default() -> Self {
+		Self([0_u8; N], PhantomData)
+	}
+}
+
+impl<const N: usize, T> PassByInner for CryptoBytes<N, T> {
+	type Inner = [u8; N];
+
+	fn into_inner(self) -> Self::Inner {
+		self.0
+	}
+
+	fn inner(&self) -> &Self::Inner {
+		&self.0
+	}
+
+	fn from_inner(inner: Self::Inner) -> Self {
+		Self(inner, PhantomData)
+	}
+}
+
+impl<const N: usize, T> PassBy for CryptoBytes<N, T> {
+	type PassBy = pass_by::Inner<Self, [u8; N]>;
+}
+
+impl<const N: usize, T> AsRef<[u8]> for CryptoBytes<N, T> {
+	fn as_ref(&self) -> &[u8] {
+		&self.0[..]
+	}
+}
+
+impl<const N: usize, T> AsMut<[u8]> for CryptoBytes<N, T> {
+	fn as_mut(&mut self) -> &mut [u8] {
+		&mut self.0[..]
+	}
+}
+
+impl<const N: usize, T> From<CryptoBytes<N, T>> for [u8; N] {
+	fn from(v: CryptoBytes<N, T>) -> [u8; N] {
+		v.0
+	}
+}
+
+impl<const N: usize, T> AsRef<[u8; N]> for CryptoBytes<N, T> {
+	fn as_ref(&self) -> &[u8; N] {
+		&self.0
+	}
+}
+
+impl<const N: usize, T> From<[u8; N]> for CryptoBytes<N, T> {
+	fn from(value: [u8; N]) -> Self {
+		Self::from_raw(value)
+	}
+}
+
+impl<const N: usize, T> TryFrom<&[u8]> for CryptoBytes<N, T> {
+	type Error = ();
+
+	fn try_from(data: &[u8]) -> Result<Self, Self::Error> {
+		if data.len() != N {
+			return Err(())
+		}
+		let mut r = [0u8; N];
+		r.copy_from_slice(data);
+		Ok(Self::from_raw(r))
+	}
+}
+
+impl<const N: usize, T> UncheckedFrom<[u8; N]> for CryptoBytes<N, T> {
```

Comment:
nit: do we need unchecked_from?

Reply:
I think that the original intent of `UncheckedFrom` was to make explicit the fact that the construction of the object is performed without actually validating the content (e.g. in case of EC based crypto that the value represents a valid point on the used curve).

While this may have sense, should be said that the same types were exposing constructors like `CryptoBytes::from_raw([u8; N])` or `ByteArray::from_raw([u8; N])` where the docs explicitly states that the check is not performed.

IMHO may be sufficient to document that the constructors from a raw byte array doesn't ensure validity of  the bytes when these are then tried to be deserialized into the higher level object.

What could go wrong if you use a bad byte array:I've not removed it yet (as other traits and redundant stuff) to keep this PR as minimal as possible
- Signatures: nothing. The verification simply fails.
- Public keys: if used to verify something. It will fail.

So, in both cases, the failure is postponed to the usage stage instead of the construction phase.

May instead have sense to remove the `From<[u8; 32]>`? And keep `UncheckedFrom` only?

Reply:
>May instead have sense to remove the From<[u8; 32]>? And keep UncheckedFrom only?

I see your point, it makes sense. Being explicit here is better. On the other hand `from` is tempting as it is shorter.

But still it is a nit, cleanup could be done in some follow-up.

Reply:
> May instead have sense to remove the `From<[u8; 32]>`? And keep `UncheckedFrom` only?

Hmmm... it could make sense, but I guess maybe only if we remove all of the other easy ways to make a `T` in an unchecked fashion?

Also, [in my refactoring PR](https://github.com/paritytech/polkadot-sdk/pull/3689) I'm using these `From` impls to convert between raw arrays when passing these types through the FFI boundary between the runtime and the host. So if we remove `UncheckedFrom` I'll have to add another wrapper strategy type just for passing these using the `UncheckedFrom`. Not a big deal, but it is extra code that will have to be added.

Reply:
I'd also like to add that the wrapped type is public. So maybe having this "UncheckedFrom" doesn't really provide a lot of safety?